mlx/backend/metal/rope.cpp

```diff
@@ -0,0 +1,56 @@
+// Copyright © 2023-2024 Apple Inc.
+
+#include <iostream>
+
+#include "mlx/backend/metal/utils.h"
+#include "mlx/extensions.h"
+#include "mlx/primitives.h"
+
+namespace mlx::core::ext {
+
+void RoPE::eval_gpu(
+    const std::vector<array>& inputs,
+    std::vector<array>& outputs) {
+  assert(inputs.size() == 1);
+  auto& in = inputs[0];
+  auto& out = outputs[0];
+
+  if (in.ndim() != 3) {
+    throw std::runtime_error(
+        "[RoPE] Only 3 dimensions are supported (batch x sequence x dims)");
+  }
+  if (dims_ != in.shape(-1)) {
+    throw std::runtime_error("[RoPE] Partial RoPE application not supported");
+  }
+  if (in.flags().row_contiguous && in.is_donatable()) {
```

Comment:
We need a contig and copy check before this right?

Reply:
Not sure what the copy check is. Also `row_contiguous` is stricter than `contiguous` is it not? ie all `row_contiguous` arrays are `contiguous` but not the other way around.

Reply:
I meant, if it's not contiguous, we should make a contiguous copy

Reply:
Actually I think I missed it, I was looking for elem to loc, but you hardcoded the strides.. so I should be ok

Reply:
Do we need to check here though that the input has the same size as the output? If it's broadcasted e.g. along the last axis it would be incorrect to donate right?

Reply:
Yeah, I hardcoded the strides cause the grid is launched with half the last dimension and it can't be delegated to a simple `elem_to_loc`. I would have to do something like multiply `pos.x` by 2 and then pass to `elem_to_loc` etc. I think this is equally readable but I am open to suggestions :-)

Regarding broadcasting, a broadcasted array wouldn't be `row_contiguous` so this check should be fine donation-wise, right?